lib/router/src/cache.rs

```diff
@@ -1,426 +1,476 @@
+/// values.
+///
+/// This contains a weak reference to a cache so that if use of the cache is
+/// dropped, it will not continue to be polled in the background.
+pub struct PurgeCache<K: Clone + Eq + Hash, V> {
+    cache: Lock<Cache<K, V>>,
+    interval: Interval,
 }
 
-/// Provides the current time within the module. Useful for testing.
-pub trait Now {
-    fn now(&self) -> Instant;
+/// Wraps a cache value so that a lock is held on the entire cache. When the
+/// access is dropped, the associated expiration time of the value in
+/// `expirations` is reset.
+pub struct Access<'a, K: Clone + Eq + Hash, V> {
+    expires: Duration,
+    expirations: &'a mut DelayQueue<K>,
+    pub(crate) node: &'a mut Node<V>,
 }
 
-/// Wraps cache values so that each tracks its last access time.
-#[derive(Debug, PartialEq)]
+/// A handle to a cache value.
 pub struct Node<T> {
-    value: T,
-    last_access: Instant,
+    key: delay_queue::Key,
+    pub(crate) value: T,
 }
 
-/// A smart pointer that updates an access time when dropped.
-///
-/// Wraps a mutable reference to a `V`-typed value.
-///
-/// When the guard is dropped, the value's `last_access` time is updated with the provided
-/// time source.
-#[derive(Debug)]
-pub struct Access<'a, T: 'a, N: Now + 'a = ()> {
-    node: &'a mut Node<T>,
-    now: &'a N,
+#[derive(Debug, PartialEq)]
+pub struct CapacityExhausted {
+    pub(crate) capacity: usize,
 }
 
-/// A handle to a `Cache` that has capacity for at least one additional value.
-#[derive(Debug)]
-pub struct Reserve<'a, K: Hash + Eq + 'a, V: 'a, N: 'a> {
+/// A handle to a cache that has capacity for at least one additional value.
+pub struct Reserve<'a, K: Clone + Eq + Hash, V> {
+    expirations: &'a mut DelayQueue<K>,
+    expires: Duration,
     vals: &'a mut IndexMap<K, Node<V>>,
-    now: &'a N,
-}
-
-#[derive(Clone, Debug, PartialEq)]
-pub struct CapacityExhausted {
-    pub capacity: usize,
 }
 
 // ===== impl Cache =====
 
-impl<K: Hash + Eq, V> Cache<K, V, ()> {
-    pub fn new(capacity: usize, max_idle_age: Duration) -> Self {
-        Self {
+impl<K: Clone + Eq + Hash, V> Cache<K, V> {
+    pub fn new(capacity: usize, expires: Duration) -> (Lock<Cache<K, V>>, PurgeCache<K, V>) {
+        assert!(capacity != 0);
+        let cache = Self {
             capacity,
+            expires,
+            expirations: DelayQueue::with_capacity(capacity),
             vals: IndexMap::default(),
-            max_idle_age,
-            now: (),
-        }
+        };
+        let cache = Lock::new(cache);
+        let bg_purge = PurgeCache {
+            cache: cache.clone(),
+            interval: Interval::new_interval(expires),
+        };
+
+        (cache, bg_purge)
+    }
+
+    pub fn access(&mut self, key: &K) -> Option<Access<K, V>> {
+        let node = self.vals.get_mut(key)?;
+        Some(Access {
+            expires: self.expires,
+            expirations: &mut self.expirations,
+            node,
+        })
```

Comment:
I think this can all be quite a bit simpler...

As it is, `lib.rs` calls this and then almost immediately drops the access, cloning the value out of the cache.  Can we get rid of all of the borrowing semantics (and the Access type entirely) and instead move the expiration-reset logic into this function? I think this would eliminate the needed for `State::Acquired`...